### PR TITLE
Fix error with --import-mode=importlib and modules containing dataclasses or pickle

### LINKED ISSUE #7856: pytest with importlib & dataclass & __future__ annotations crash
I'm transitioning from pytest regular imports to `--import-mode=importlib` imports due to namespace collisions in tests. And I'm getting the following crash, no plugins installed, `pytest 6.1.1` `CPython 3.8.6`, the issue reproduces on Mac & Linux.

File names test_xx.py:
```python
from __future__ import annotations

from dataclasses import dataclass

@dataclass
class TestHelper:
    value: str
```

running `pytest --import-mode=importlib test_xx.py` and I get the following crash:
```
==================================================== test session starts ====================================================
platform darwin -- Python 3.8.5, pytest-6.1.1, py-1.9.0, pluggy-0.13.1
rootdir: /private/tmp/qwer/dir
collected 0 items / 1 error

========================================================== ERRORS ===========================================================
________________________________________________ ERROR collecting test_xx.py ________________________________________________
test_xx.py:6: in <module>
    class TestHelper:
/usr/local/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/dataclasses.py:1019: in dataclass
    return wrap(cls)
/usr/local/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/dataclasses.py:1011: in wrap
    return _process_class(cls, init, repr, eq, order, unsafe_hash, frozen)
/usr/local/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/dataclasses.py:861: in _process_class
    cls_fields = [_get_field(cls, name, type)
/usr/local/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/dataclasses.py:861: in <listcomp>
    cls_fields = [_get_field(cls, name, type)
/usr/local/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/dataclasses.py:712: in _get_field
    and _is_type(f.type, cls, typing, typing.ClassVar,
/usr/local/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/dataclasses.py:658: in _is_type
    ns = sys.modules.get(cls.__module__).__dict__
E   AttributeError: 'NoneType' object has no attribute '__dict__'
================================================== short test summary info ==================================================
ERROR test_xx.py - AttributeError: 'NoneType' object has no attribute '__dict__'
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
===================================================== 1 error in 0.28s ======================================================
```

pip list:
```
Package    Version
---------- -------
attrs      20.2.0
iniconfig  1.0.1
packaging  20.4
pip        20.1.1
pluggy     0.13.1
py         1.9.0
pyparsing  2.4.7
pytest     6.1.1
setuptools 50.3.0
six        1.15.0
toml       0.10.1
wheel      0.35.1
```

### LINKED ISSUE #7859: pytest importlib doesn't work with pickle
I'm transitioning from pytest regular imports to --import-mode=importlib imports due to namespace collisions in tests. And I'm getting the following crash, no plugins installed, `pytest 6.1.1` `CPython 3.8.6`, the issue reproduces on Mac & Linux.

Create a file named test_xx.py:

```python
import pickle

def do_action():
    pass

def test_x():
    pickle.dumps(do_action)
```

```
$ pytest -v --import-mode=importlib test_xx.py
==================================================== test session starts ====================================================
platform darwin -- Python 3.8.5, pytest-6.1.1, py-1.9.0, pluggy-0.13.1 -- /Users/yoav/.local/share/virtualenvs/qwer-HyU7-4uu/bin/python
cachedir: .pytest_cache
rootdir: /private/tmp/qwer
collected 1 item

test_xx.py::test_x FAILED                                                                                             [100%]

========================================================= FAILURES ==========================================================
__________________________________________________________ test_x ___________________________________________________________

    def test_x():
>       pickle.dumps(do_action)
E       _pickle.PicklingError: Can't pickle <function do_action at 0x10f8aa3a0>: import of module 'test_xx' failed

test_xx.py:7: PicklingError
================================================== short test summary info ==================================================
FAILED test_xx.py::test_x - _pickle.PicklingError: Can't pickle <function do_action at 0x10f8aa3a0>: import of module 'tes...
===================================================== 1 failed in 0.05s =====================================================
```

Running the same code without pytest or with pytest without importlib work as expected.

pip list:

```
Package    Version
---------- -------
attrs      20.2.0
iniconfig  1.0.1
packaging  20.4
pip        20.1.1
pluggy     0.13.1
py         1.9.0
pyparsing  2.4.7
pytest     6.1.1
setuptools 50.3.0
six        1.15.0
toml       0.10.1
wheel      0.35.1
```

I encountered this issue because we have tests using `multiprocessing.Pool`, and they fail on pickling the target function.

### PR DESCRIPTION
**Pull request by tadeu**
fixes #7856, fixes #7859

### CODE REVIEW

File: src/_pytest/pathlib.py
Comment (nicoddemus) [inline src/_pytest/pathlib.py:483]: Unfortunately AFAIK one of the "advantages" of using `importlib` was that it does not change `sys.path` or `sys.modules` at all... or perhaps the main advantage is actually not changing `sys.path` and updating `sys.modules` is fine?

@asottile what do you think? Also @The-Compiler which has some experience in this matter.

**Comment by nicoddemus** [comment]
Thanks a lot @tadeu!

I'm not 100% sure the fix is correct (or even if it is fixable), but let see what the other maintainers think. 

I might have the wrong idea about what benefits `importlib` should bring to the table: I imagined it was not changing `sys.path` and `sys.modules`, but I might be wrong about the latter.

File: src/_pytest/pathlib.py
Comment (tadeu) [reply src/_pytest/pathlib.py:483]: I hope it's the latter because some code out there really depend on the module being in `sys.modules`, e.g., `ns = sys.modules.get(cls.__module__).__dict__` in `dataclasses.py`.

Comment (The-Compiler) [reply src/_pytest/pathlib.py:483]: I think we should strive to be as close as possible to a "normal" import in order to not break people's expectations. We definitely should avoid messing with `sys.path` (as this will change how other imports work!), but I think we should add our module to `sys.modules` so that our imports aren't exotic in any way.

The `importlib.import_module` convenience method also does that:

```pycon
>>> import importlib
>>> 'pytest' in sys.modules
False
>>> importlib.import_module('pytest')
<module 'pytest' from '/usr/lib/python3.8/site-packages/pytest/__init__.py'>
>>> sys.modules['pytest']
<module 'pytest' from '/usr/lib/python3.8/site-packages/pytest/__init__.py'>
```

Which makes me wonder - why are we hand-rolling everything here based on the more low-level functions rather than just using the convenience function? The importlib doc also has a ["Importing a source file directly"](https://docs.python.org/3/library/importlib.html#importing-a-source-file-directly) recipe which looks very similar to what we do, and also adds the module to `sys.modules`.

Comment (nicoddemus) [reply src/_pytest/pathlib.py:483]: > I hope it's the latter because some code out there really depend on the module being in sys.modules, e.g., ns = sys.modules.get(cls.__module__).__dict__ in dataclasses.py.

Definitely! `importmode=importlib` definitely doesn't cover all cases, at least currently.

> If pytest really depend on modules not being added to sys.modules when using importlib, perhaps a test should be added for this behavior?

It is not that pytest depends on that, is how other modules interact with that and how that module is added to `sys.modules` (see below).

> Also, if it's the case, maybe the module could be added temporarily to sys.modules, just for spec.loader.exec_module(mod).

I had the same idea a few years back, but @RonnyPfannschmidt correctly pointed out that removing a module from `sys.modules` might cause havoc later (think local imports failing when the normal import succeeded).

> The importlib doc also has a "Importing a source file directly" recipe which looks very similar to what we do, and also adds the module to sys.modules.

Indeed. 

Just to recap, one of the reason this option exists is to allow different test modules with the same name, without a package, to be imported. This is the classic problem where the test hierarchy below fails:

```
tests/
  a/
    test_foo.py
  b/
    test_foo.py
```

(Without packages)

By the time we get to `b/test_foo.py`, under normal import mode, we change `sys.path` to point to `tests/b`, and try to import `test_foo.py`; but then we realize `test_foo` is already in `sys.modules`, but point to a different file, and raise a module mismatch error.

With `--importmode=importlib` the above works, because to import both `test_foo.py` files we don't change `sys.path` or `sys.modules`, so there's no conflict. But of course, having a module object which is not in `sys.modules` breaks people's expectations.

Now if we want to add both `test_foo` modules in `sys.modules`, which key should we use to avoid conflicts? Note that we can't rely on `sys.path` at all (as we don't want to change it). Perhaps use the full file path, replacing `os.path.sep` by `.`, would resolve this?

Full disclosure: I'm just stating how things work and the rationale behind it. I'm more than happy to improve this. 😁

Comment (RonnyPfannschmidt) [reply src/_pytest/pathlib.py:483]: - This brings back the topic of import conflicts and requires some discussion and crosscheck that i can't do soon

Comment (asottile) [reply src/_pytest/pathlib.py:483]: I'm late to the party sorry was very busy today :)

the points above are what make this tricky: the `importlib` mode was designed to avoid module namespace conflicts and the keys in `sys.modules` would be those same conflicts (so we'd need to come up with some unique thing that *also* gets set as `__name__` in that module)

Comment (RonnyPfannschmidt) [reply src/_pytest/pathlib.py:483]: The general idea i have for that is a general concept of collection roots that either belong to a existing package, or get a otherwise unique name

Unfortunately i don't have the capacity to work on this soon

Comment (nicoddemus) [reply src/_pytest/pathlib.py:483]: How about using the module name relative to `config.root`, perhaps adding a uniquely-named prefix? I suppose that would produce unique names and also something one would look and recognize.

So in the usual project:

```
src/
tests/
pytest.ini
```

We can use `__pytest_importmode_importlib__.tests.a.test_foo`, I doubt this would clash with anything (just using `tests.a.test_foo` was my initial idea, but this can clash depending on the project layout and `config.root`).

I believe this module name can be considered experimental/internal detail (and warn that in the docs as such) until we are sure it works or come up with a better idea.

Comment (tadeu) [reply src/_pytest/pathlib.py:483]: I've done some manual testing (comparing pytest against importing things manually, also testing running pytest with multiple roots). Note that even though some modules with the same name won't appear in `sys.modules`, everything already seems to work fine (with the changes in this PR), even with `conftest` files. (Maybe there's some overlooked use case?)

It would probably be problematic if tests tried to import other test files, but I think that they are not supposed to do that, specially with implicit namespace packages.

The tests can be found here: https://github.com/tadeu/_pytest-7870

Diff between running "manually" and with pytest is here: https://github.com/tadeu/_pytest-7870/blob/main/diff1.diff

I'd be happy to transform that into unit tests in pytest/testing if you think it's worthy.

**Comment by nicoddemus** [comment]
(Marking this with the `hacktoberfest-accepted` label as this is clearly a quality contribution, but I'm afraid it is complex and might still require a bunch of discussions to move forward)

File: src/_pytest/pathlib.py
Comment (nicoddemus) [reply src/_pytest/pathlib.py:483]: Still overwriting the modules will definitely prove problematic in real world test suites... pytest is used by thousands of projects, I'm sure this will break some people.

I think the proper way to handle this is to use a unique key to insert the module in `sys.modules`.

Comment (tadeu) [reply src/_pytest/pathlib.py:483]: > We can use `__pytest_importmode_importlib__.tests.a.test_foo`, I doubt this would clash with anything (using only `tests.a.test_foo` was my initial idea, but this can clash depending on the project layout and `config.root`).

After the discussions I can also try to implement this, I hope it's not too much of a rabbit hole in the internals 😁🐰

Comment (nicoddemus) [reply src/_pytest/pathlib.py:483]: Thanks! Let's see what others think first though. 👍

Comment (The-Compiler) [reply src/_pytest/pathlib.py:483]: If that works with dataclasses, pickle, or the other scenarios mentioned in #7245, I'd say go for it? I think we don't necessarily need to get *everything* right as part of this PR. It's clearly an improvement of the status quo (where pytest breaks with a variety of stdlib/builtin functionality).

Given that `--import-mode=importlib` isn't the default yet and we're asking for feedback with it, I think it's okay to have slightly breaking changes in the future (before it's made the default) rather than stalling this PR further because we don't actually know what's the right approach.

Comment (nicoddemus) [reply src/_pytest/pathlib.py:483]: I agree with @The-Compiler's reasoning.

What others think? Can we give green light to @tadeu to implement the idea of using a unique key based on `config.root`? Other ideas?

Comment (nicoddemus) [reply src/_pytest/pathlib.py:483]: Anybody? It would be interesting to get this kicking again.

**Comment by nicoddemus** [comment]
Picked this up again, I've rebased and changed the approach to use a dotted name for modules, anchored to `config.rootpath` to obtain a unique name.

**Review by bluetech — COMMENTED** [review]
I haven't reviewed in depth but LGTM.

One thing I wonder is, can the empty modules added by `insert_missing_modules` cause any trouble? If a module name already exists, it doesn't override it, which is good, but I wonder about the opposite case, when `insert_modules` inserts an empty module, and *after* that a real module with that name wants to be imported. Is this case possible? And if so, what will happen?

**Comment by nicoddemus** [comment]
> If a module name already exists, it doesn't override it, which is good, but I wonder about the opposite case, when insert_modules inserts an empty module, and after that a real module with that name wants to be imported. Is this case possible? And if so, what will happen?

Hmm good points, I think it will blow up, as someone might be expecting the real module (to say, call a function from it) and they will get a fake module instead.

However I think this is highly unlikely, because if one is using `importlib` mode, it is a premise that they won't be able to import things from other test modules, so I'm having a hard time coming up with a plausible example where this might be an issue.

**Comment by bluetech** [comment]
Got it. If it turns out to be a problem in practice, then worst case we can revert or fix it some other way.

**Review by bluetech — APPROVED** [review]

**Comment by nicoddemus** [comment]
Rebased on latest `main`, needs to be squashed on merge.

Sorry for dropping the ball on this one, this totally got out of my radar. 

I will merge it Tomorrow unless someone wants more time to give it another pass, in which case just let me know here.

**Pull request merged**
